src/lib.rs

```diff
@@ -271,96 +280,165 @@ fn filter_attr(attr: &&syn::Attribute) -> bool {
-            trace!("Skipping setter.");
-        }
+        setter_fns.push(derive_setter(f, &f_opts, &attrs));
+        builder_fields.push(derive_builder_field(f, &f_opts, &attrs));
+        initializers.push(derive_initializer(f, &f_opts));
+        // NOTE: Looking forward for computation in interpolation
+        // - https://github.com/dtolnay/quote/issues/10
+        // => `quote!(#{f.vis} ...)
+    }
 
-        tokens
-    });
+    let builder_vis = opts.builder_visibility();
+    let build_fn = {
+        let ref_self = match *opts.field_defaults().setter_pattern() {
+            SetterPattern::Owned => quote!(self),
+            SetterPattern::Mutable => quote!(&self),
+            SetterPattern::Immutable => quote!(&self),
+        };
+        quote!(
+            #builder_vis fn build(#ref_self) -> Result<#struct_name #ty_generics, String> {
+                Ok(#struct_name{
+                    #(#initializers)*
+                })
+            }
+        )
+    };
 
+    // We need to `#[derive(Clone)]` only for the immutable builder pattern
     quote! {
+        #[derive(Default, Clone)]
+        #builder_vis struct #builder_name #ty_generics #where_clause {
+            #(#builder_fields)*
+        }
+
         #[allow(dead_code)]
-        impl #impl_generics #name #ty_generics #where_clause {
-            #(#funcs)*
+        impl #impl_generics #builder_name #ty_generics #where_clause {
+            #(#setter_fns)*
+
+            #build_fn
         }
     }
 }
 
-fn derive_setter(f: &syn::Field, opts: &Options) -> quote::Tokens {
-    trace!("Deriving setter.");
-    let ty = &f.ty;
+fn derive_builder_field(f: &syn::Field, opts: &FieldOptions, attrs: &Vec<&syn::Attribute>) -> quote::Tokens {
+    if opts.setter_enabled() {
+        trace!("Deriving builder field for {}.", opts.field_name());
+        let vis = &f.vis;
+        let ident = &f.ident;
+        let ty = &f.ty;
+        quote!(#(#attrs)* #vis #ident: Option<#ty>,)
+    } else {
+        trace!("Skipping builder field for {}.", opts.field_name());
+        quote!()
+    }
+}
+
+fn derive_initializer(f: &syn::Field, opts: &FieldOptions) -> quote::Tokens {
+    trace!("Deriving initializer for {}.", opts.field_name());
+
+    let err_uninitizalied = format!("{} must be initialized", opts.field_name());
     let pattern = opts.setter_pattern();
-    let vis = opts.setter_visibility();
-    let fieldname = f.ident.as_ref().expect(&format!("Missing identifier for field {:?}.", f));
-    let funcname = if opts.setter_prefix().len() > 0 {
-        Cow::Owned(syn::Ident::new(format!("{}_{}", opts.setter_prefix(), fieldname)))
+    let ident = &f.ident;
+    // let attrs = &f.attrs;
+
+    if opts.setter_enabled() {
+        let initializer = match *pattern {
+            SetterPattern::Owned => quote!(
+                    #ident: self.#ident.ok_or(#err_uninitizalied)?,
+                ),
+            SetterPattern::Mutable
+            | SetterPattern::Immutable => quote!(
```

Comment:
I think it's actually more common on the end of the line.

But I thought it would make more sense at the beginning because when you say "A or B" you would put a pause like this "A ... or B" and not "A or ... B". But let's go with the crowd.